internal/controllers/machine/machine_controller_noderef_test.go

```diff
@@ -495,35 +687,42 @@ func TestReconcileNodeTaints(t *testing.T) {
 					},
 				},
 			},
-		},
-		{
-			name: "if the node does not have the uninitialized taint it should remain absent from the node",
-			node: &corev1.Node{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: "test-node-1",
-				},
-				Spec: corev1.NodeSpec{
-					Taints: []corev1.Taint{},
+			expectedAnnotations: map[string]string{
+				clusterv1.LabelsFromMachineAnnotation: "",
+			},
+			expectedTaints: []corev1.Taint{
+				{
+					Key:    "node-role.kubernetes.io/control-plane",
+					Effect: corev1.TaintEffectNoSchedule,
 				},
+				{Key: "node.kubernetes.io/not-ready", Effect: "NoSchedule"}, // Added by the API server
 			},
 		},
 	}
 
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
+	r := Reconciler{Client: env}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
 			g := NewWithT(t)
-			fakeClient := fake.NewClientBuilder().WithObjects(tt.node).WithScheme(fakeScheme).Build()
-			nodeBefore := tt.node.DeepCopy()
-			r := &Reconciler{}
-			g.Expect(r.reconcileNodeTaints(ctx, fakeClient, tt.node)).Should(Succeed())
-			// Verify the NodeUninitializedTaint is dropped.
-			g.Expect(tt.node.Spec.Taints).ShouldNot(ContainElement(clusterv1.NodeUninitializedTaint))
-			// Verify all other taints are same.
-			for _, taint := range nodeBefore.Spec.Taints {
-				if !taint.MatchTaint(&clusterv1.NodeUninitializedTaint) {
-					g.Expect(tt.node.Spec.Taints).Should(ContainElement(taint))
-				}
-			}
+			oldNode := tc.oldNode.DeepCopy()
+
+			g.Expect(env.Create(ctx, oldNode)).To(Succeed())
+			t.Cleanup(func() {
+				_ = env.Cleanup(ctx, oldNode)
+			})
+
+			err := r.patchNode(ctx, oldNode, tc.newLabels, tc.newAnnotations)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			g.Eventually(func(g Gomega) {
+				gotNode := &corev1.Node{}
+				err = env.Get(ctx, client.ObjectKeyFromObject(oldNode), gotNode)
+				g.Expect(err).ToNot(HaveOccurred())
+
+				g.Expect(gotNode.Labels).To(Equal(tc.expectedLabels), cmp.Diff(gotNode.Labels, tc.expectedLabels))
```

Comment:
```suggestion
				g.Expect(gotNode.Labels).To(BeComparableTo(tc.expectedLabels))
```
nit: How about this for simplicity? `BeComparableTo` internally does the same and prints a `cmp.Diff` if there is a difference. 

(same at other 2 places)